File: apps/android/app/src/test/java/com/litter/android/state/TerminalOutputBufferTests.kt
package com.litter.android.state

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class TerminalOutputBufferTests {
    @Test
    fun `flush batches output and preserves order`() {
        val buffer = TerminalOutputBuffer(maxChars = 128, maxPendingChars = 128)

        buffer.append("hello ".toByteArray())
        buffer.append("world".toByteArray())

        assertTrue(buffer.hasPending)
        assertTrue(buffer.flush())
        assertEquals("hello world", buffer.text)
        assertFalse(buffer.hasPending)
    }

    @Test
    fun `flush keeps bounded tail for burst output`() {
        val buffer = TerminalOutputBuffer(maxChars = 10, maxPendingChars = 20)

        buffer.append("0123456789".toByteArray())
        buffer.flush()
        buffer.append("abcdefghij".toByteArray())
        buffer.flush()

        assertEquals("abcdefghij", buffer.text)
    }

    @Test
    fun `pending output is bounded before scheduled flush`() {
        val buffer = TerminalOutputBuffer(maxChars = 64, maxPendingChars = 6)

        buffer.append("abcdef".toByteArray())
        buffer.append("ghijkl".toByteArray())
        buffer.flush()

        assertEquals("ghijkl", buffer.text)
    }
}

File: apps/ios/Tests/LitterTests/TerminalOutputBufferTests.swift
import XCTest
@testable import Litter

final class TerminalOutputBufferTests: XCTestCase {
    func testFlushBatchesOutputAndPreservesOrder() {
        var buffer = TerminalOutputBuffer(maxCharacters: 128, maxPendingCharacters: 128)

        buffer.append(Data("hello ".utf8))
        buffer.append(Data("world".utf8))

        XCTAssertTrue(buffer.hasPendingOutput)
        XCTAssertTrue(buffer.flush())
        XCTAssertEqual(buffer.text, "hello world")
        XCTAssertFalse(buffer.hasPendingOutput)
    }

    func testFlushKeepsBoundedTailForBurstOutput() {
        var buffer = TerminalOutputBuffer(maxCharacters: 10, maxPendingCharacters: 20)

        buffer.append(Data("0123456789".utf8))
        _ = buffer.flush()
        buffer.append(Data("abcdefghij".utf8))
        _ = buffer.flush()

        XCTAssertEqual(buffer.text, "abcdefghij")
    }

    func testPendingOutputIsBoundedBeforeScheduledFlush() {
        var buffer = TerminalOutputBuffer(maxCharacters: 64, maxPendingCharacters: 6)

        buffer.append(Data("abcdef".utf8))
        buffer.append(Data("ghijkl".utf8))
        _ = buffer.flush()

        XCTAssertEqual(buffer.text, "ghijkl")
    }
}
